WordPress/Classes/Services/SuggestionService.swift

```diff
@@ -0,0 +1,118 @@
+import Foundation
+
+extension NSNotification.Name {
+    static let suggestionListUpdated = NSNotification.Name("SuggestionListUpdated")
+}
+
+@objc extension NSNotification {
+    public static let suggestionListUpdated = NSNotification.Name.suggestionListUpdated
+}
+
+@objc public class SuggestionService: NSObject {
+
+    // NSCache works with classes such as NSArray, not structs such as Suggestion or [Suggestion].
+    private class Wrapper<T>: NSObject {
+        let value: T
+        init(_ value: T) {
+            self.value = value
+        }
+    }
+
+    private let suggestionsCache = NSCache<NSNumber, Wrapper<[Suggestion]>>()
+    private var siteIDsCurrentlyBeingRequested = [NSNumber]()
+
+    private static let shared = SuggestionService()
+
+    @objc static func sharedInstance() -> SuggestionService {
+        return shared
+    }
+
+    private override init() {}
+
+    /**
+    Returns the cached @mention suggestions (if any) for a given siteID.  Calls
+    updateSuggestionsForSiteID if no suggestions for the site have been cached.
+
+    @param siteID ID of the blog/site to retrieve suggestions for
+    @return An array of suggestions
+    */
+    @objc func suggestions(for siteID: NSNumber) -> [Suggestion]? {
+        if let cachedSuggestions = suggestionsCache.object(forKey: siteID) {
+            return cachedSuggestions.value
+        }
+        updateSuggestions(for: siteID)
+        return nil
+    }
+
+    /**
+    Performs a REST API request for the siteID given.
+
+    @param siteID ID of the blog/site to retrieve suggestions for
+    */
+    private func updateSuggestions(for siteID: NSNumber) {
+
+        // if there is already a request in place for this siteID, just wait
+        guard !siteIDsCurrentlyBeingRequested.contains(siteID) else { return }
+
+        // add this siteID to currently being requested list
+        siteIDsCurrentlyBeingRequested.append(siteID)
+
+        let suggestPath = "rest/v1.1/users/suggest"
+        let context = ContextManager.shared.mainContext
+        let accountService = AccountService(managedObjectContext: context)
+        let defaultAccount = accountService.defaultWordPressComAccount()
+        let params = ["site_id": siteID]
+
+        defaultAccount?.wordPressComRestApi.GET(suggestPath, parameters: params, success: { [weak self] responseObject, httpResponse in
+            guard let `self` = self else { return }
+            guard let payload = responseObject as? [String: Any] else { return }
+            guard let restSuggestions = payload["suggestions"] as? [[String: Any]] else { return }
+
+            let suggestions = restSuggestions.map { Suggestion(dictionary: $0) }
+            self.suggestionsCache.setObject(Wrapper(suggestions), forKey: siteID)
```

Comment:
The original code sets a cost for the suggestions based on their count. Is that no longer required? (I think probably unnecessary)

Reply:
I'd overlooked that, thanks for catching!

Based on the docs for `setObject(_:forKey:cost:)`, it does seem unnecessary as you suggest (emphasis mine):

> Pass in 0 for the cost value if you otherwise have nothing useful to pass, **or simply use the setObject:forKey: method, which does not require a cost value to be passed in**.

So I'd agree this change is ok to include here in this refactor.